### PR TITLE
Add oatpp package to fix #1228

### LINKED ISSUE #1228: [request] oatpp/1.0.0
### Package Details
  * Package Name/Version: **oatpp/1.0.0**
  * Website: **https://oatpp.io/**
  * Source code: **https://github.com/oatpp/oatpp**

### Description Of The Library / Tool
Modern Web Framework for C++. High performance, simple API, cross platform, zero dependency.

### PR DESCRIPTION
**Pull request by frek818**
Specify library name and version:  oatpp/1.0.0

- [*] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [*] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [*] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [*] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

close https://github.com/conan-io/conan-center-index/issues/1228

### CODE REVIEW

**Comment by uilianries** [comment]
@frek818 please, sign the CLA

File: recipes/oatpp/all/test_package/example.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "oatpp/core/concurrency/SpinLock.hpp"
+#include "oatpp/core/base/Environment.hpp"
+
+#include <iostream>
+#include <atomic>
+
+
+void runTests() {
+    OATPP_RUN_TEST(oatpp::test::network::UrlTest);
```
Comment (lganzzzo) [inline recipes/oatpp/all/test_package/example.cpp:11]: Hey @frek818 ,

I recommend adding the following two tests (as they can fail on a `shared` lib build):
- https://github.com/oatpp/oatpp/blob/master/test/oatpp/core/data/mapping/type/TypeTest.hpp
- https://github.com/oatpp/oatpp/blob/master/test/oatpp/parser/json/mapping/DeserializerTest.hpp

**Comment by uilianries** [comment]
Please, try to keep it simple. The test package, as the name suggests, it's for testing your package, not an unit test, you only need to check if the package has been built correctly, not running all unit test to check the project behavior. I would to see only one file for test package if possible.

**Comment by lganzzzo** [comment]
Let's then have the [DeserializerTest](https://github.com/oatpp/oatpp/blob/master/test/oatpp/parser/json/mapping/DeserializerTest.hpp) only. 
It should be enough

File: recipes/oatpp/all/test_package/example.cpp (hunk 1 above)
Comment (frek818) [reply recipes/oatpp/all/test_package/example.cpp]: @uilianries and @lganzzzo  or Is the best practice to create a separate package for the extension modules (e.g. oatpp-websocket, oatpp-swagger) and require oatpp/1.0.0 for each extension?

I start working on a PR for oatpp-websocket (https://github.com/conan-io/conan-center-index/pull/1689). I think an alternative option would be to add options to the oatpp recipe that would embed the extension in this packge.

**Comment by frek818** [comment]
> Some configurations of 'oatpp/1.0.0' failed in build 8 (`c5b656211fc36cdbe492d7d7d8a5b3c570ce1fcb`):
> 
> * **Windows x86_64, Debug, Visual Studio 14,  MTd. Options:  oatpp:shared-False**
>   
>   * [Log download](https://c3i.jfrog.io/c3i/misc/logs/pr/1677/8/oatpp/1.0.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stdout.txt)
>   * [Log error Download](https://c3i.jfrog.io/c3i/misc/logs/pr/1677/8/oatpp/1.0.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/create_stderr.txt)
>   * [Profile](https://c3i.jfrog.io/c3i/misc/logs/pr/1677/8/oatpp/1.0.0/f27dcd2d5e453cb01f9324a959461903ce8116a0/profile.txt)
> * [Access to all the logs](https://c3i.jfrog.io/c3i/misc/logs/pr/1677/8/oatpp/1.0.0/)

@lganzzzo, Do you know know what the issue is here? I've only test the package on macos and linux. I don't have a window machne to test this on.

**Comment by lganzzzo** [comment]
@frek818 ,

I'm not sure what is the last error about.
`048d6cda82e46f20d092bf9aa7a929b7b056e7c3`
```bash
LINK : fatal error LNK1104: cannot open file '..\..\lib\oatpp.lib'
```
Looks like for other VS versions it was built successfully - [VS-16 build](https://c3i.jfrog.io/c3i/misc/logs/pr/1677/9/oatpp/1.0.0/6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7/)

**Googling**: The MS documentation suggests that it's something wrong with the path library path, others claim that they were able to fix it by just re-running the build.

@uilianries do you have any ideas/suggestions?

**Comment by uilianries** [comment]
@frek818 oatpp doesn't export its symbols on Windows, it produces DLLs but you need to use dynamic methods load. 

My suggestion is CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS, but asking to the upstream to export the symbols in a next release. If it doesn't work, you can drop the shared option on Windows:

    def configure(self):
        if self.settings.os == "Windows" and self.options.shared:
            raise ConanInvalidConfiguration("oatpp can not be built as shared library on Windows")

**Comment by lganzzzo** [comment]
Hey @frek818 ,

The `5ea5d2f06ec3d9767eabd30fc601b615da93c8a4` - Linux x86_64, Release, gcc 4.9, libstdc++

It's a known issue - (*from oatpp dev chat*) - `because put_time manipulator is missing in gcc 4.8 and 4.9`.

It may be fixed in later versions by changing `put_time` to `strftime`.

For now, I suggest disabling `gcc 4.8` and `4.9` builds.

Regards,
Leonid

File: recipes/oatpp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename("oatpp-{0}".format(self.version), self._source_subfolder)
+
+    def configure(self):
+        if self.settings.os == "Windows" and self.options.shared:
+            raise ConanInvalidConfiguration("oatpp can not be built as shared library on Windows")
+
+        if self.settings.compiler == "gcc" and self.settings.compiler.version in ["4.8", "4.9"]:
+            raise ConanInvalidConfiguration("oatpp can not be built with gcc 4.8 or 4.9")
```
Comment (uilianries) [inline recipes/oatpp/all/conanfile.py:37]:
```suggestion
        if self.settings.compiler == "gcc" and tools.Version(self.settings.compiler.version) < "5":
            raise ConanInvalidConfiguration("oatpp requires GCC >=5")
```

Comment (frek818) [reply recipes/oatpp/all/conanfile.py:37]: Conan has some nifty APIs. Thanks for the suggestion.

**Pull request merged**